### PR TITLE
ioBundle->NetworkPortStatus matching tightening for IsPort()

### PR DESCRIPTION
**Pull request by andrewd-zededa**
Check all three identifiers: ifname, usbaddr, pciaddr
Some io types can have an empty string in one of these fields.
Expand matching conditions to ensure ioBundles are not incorrectly matched to a different NetworkPort.

While modifying the parameters to types.IsPort() I discovered an unused isPort() which referenced a lock that appears to be used nowhere else and removed both.

### CODE REVIEW

File: pkg/pillar/types/dns.go [hunk 1]
```diff
@@ -683,6 +683,9 @@ func getLocalAddrListImpl(dns DeviceNetworkStatus,
 
 // Check if an interface name is a port owned by nim
 func IsPort(dns DeviceNetworkStatus, ifname string) bool {
+	if ifname == "" {
+		return false
```
Comment (milan-zededa) [inline pkg/pillar/types/dns.go:687]: We cannot make this assumption since we allow cellular modems to have empty interface name in the device model: https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/zedagent/parseconfig.go#L1137-L1162

Comment (eriknordmark) [reply pkg/pillar/types/dns.go:687]: @milan-zededa but do we want empty ifname to match all/any port with an empty ifname in that case?

Comment (milan-zededa) [reply pkg/pillar/types/dns.go:687]: No, the fix for `IsPort` (to handle wwan adapter with undefined interface name) could be to include PCI and USB addresses from the IoBundle as arguments, and match NetworkPortStatus by non-empty addresses.

Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go]: pushed up a change to incorporate @milan-zededa suggestion.

Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go]: Thanks @milan-zededa

**Review by eriknordmark — COMMENTED** [review]
LGTM but are there other functions which compare on ifname which should also compare on all three fields?

File: pkg/pillar/types/dns.go [hunk 2]
```diff
@@ -31,6 +31,8 @@ type DeviceNetworkStatus struct {
 
 type NetworkPortStatus struct {
 	IfName       string
+	USBAddr      string
```
Comment (christoph-zededa) [inline pkg/pillar/types/dns.go:34]: I think if you add `USBAddr` then you should also add `USBProd`.

**Comment by andrewd-zededa** [comment]
> LGTM but are there other functions which compare on ifname which should also compare on all three fields?

@eriknordmark there are a series of other helper functions in types/dns.go (possibly elsewhere too): IsL3Port(), IsMgmtPort(), GetPortCost(), ...and all of these appear to only check Ifname for equivalence of the device.  At the moment I'm not sure how many of these are relevant for device passthrough specifically..

File: pkg/pillar/types/dns.go (hunk 2 above)
Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go:34]: @christoph-zededa NetworkPortConfig doesn't currently have a USBProd field.  Is USBAddr enough of an identifier?

Comment (eriknordmark) [reply pkg/pillar/types/dns.go:34]: I also think it makes sense to add that field since we allow specifying only the USBProd for USB devices.

**Comment by eriknordmark** [comment]
> 
> @eriknordmark there are a series of other helper functions in types/dns.go (possibly elsewhere too): IsL3Port(), IsMgmtPort(), GetPortCost(), ...and all of these appear to only check Ifname for equivalence of the device. At the moment I'm not sure how many of these are relevant for device passthrough specifically.

> Maybe this should be split out to some form of a IsSameDevice() call (or some better name).

You'd actually need to create some Lookup function if you want to break out the comparison of the different identifiers.

Looking at the functions you listed:
1. IsMgmtPort() is only called for ifnames which are in DeviceNetworkStatus.Ports. But based on Milan's comment we can have modems which do not have an ifname, so makes sense to add the args there.
2. IsL3Port() is no longer used hence can be deleted.
3. GetPortCost() has the same considerations as IsMgmtPort.

**Comment by eriknordmark** [comment]
But I don't understand if we can have empty ifname arguments to these functions. There are other functions in dns.go which build a list of ifnames and then use that to get a list of IP addresses, and that will produce the wrong thing if a modem interface without an ifname is an actual Port which has an IP address. So I'm a bit confused - need to re-read Milan's earlier comment.

File: pkg/pillar/types/dns.go (hunk 2 above)
Comment (christoph-zededa) [reply pkg/pillar/types/dns.go:34]: F.e. we might use a USB ethernet adapter and specify it in the model with vendor id and product id.

Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go:34]: Ok i'll make the changes

**Comment by andrewd-zededa** [comment]
@eriknordmark Milan's comment references propagatePhyioAttrsToPort() which calls handleMissingIfname() so the NetworkPortConfig seems like it should be safe but somehow NetworkPortStatus had an empty field.

File: pkg/pillar/types/dns.go (hunk 2 above)
Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go:34]: @christoph-zededa @eriknordmark pushed up changes plumbing the USBProd field up through NetworkPortConfig and NetworkPortStatus

**Comment by andrewd-zededa** [comment]
Latest pushes include USBProd plumbing through NetworkPortConfig/Status removal of unused IsL3Port() and spreading the expanded check of usbaddr, usbprod, pciaddr to IsMgmtPort() and GetPortCost()

File: pkg/pillar/types/dns.go [hunk 3]
```diff
@@ -682,33 +685,40 @@ func getLocalAddrListImpl(dns DeviceNetworkStatus,
 }
 
 // Check if an interface name is a port owned by nim
-func IsPort(dns DeviceNetworkStatus, ifname string) bool {
+func IsPort(dns DeviceNetworkStatus, ifname string, usbaddr string, usbprod string, pciaddr string) bool {
 	for _, us := range dns.Ports {
 		if us.IfName != ifname {
```
Comment (milan-zededa) [inline pkg/pillar/types/dns.go:690]: @andrewd-zededa As additional fix, we might need to apply this check only if ifname is non-empty, i.e.
```
if ifname != "" && us.IfName != ifname {
  continue
}
```
Even when wwan adapter has empty ifname in the model (because PCI or USB address is used instead to match it), DPCManager will still report the interface name in DeviceNetworkStatus once it is [discovered](https://github.com/lf-edge/eve/blob/master/pkg/pillar/dpcmanager/wwan.go#L134). (we need interface name for metrics etc.)
So to avoid returning false for wwan adapter with empty ifname in the model but actually used by NIM, match by ifname only if non empty.

Otherwise the PR LGTM.

Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go]: @milan-zededa ok modified all the ifname compares to also check if not empty, in the latest push.

**Comment by andrewd-zededa** [comment]
@eriknordmark @milan-zededa when this is approved/merged I need to backport to one or more stable/lts branches.

**Review by rene — APPROVED** [review]
LGTM... kicking off tests...

**Comment by rucoder** [comment]
> LGTM... kicking off tests...

@rene still not clear whether @andrewd-zededa has tested on real device

**Comment by rene** [comment]
> > LGTM... kicking off tests...
> 
> @rene still not clear whether @andrewd-zededa has tested on real device

ok, let's run the tests at least, he still needs to update the commit message, so when he pushes the final version we can review it again....

**Comment by andrewd-zededa** [comment]
@rene updated commit message with your requests

**Review by OhmSpectator — APPROVED** [review]

**Review by eriknordmark — APPROVED** [review]
I don't know if this is all we need - I'm concerned that we do not set ifname for all WWAN ports which would make them not be used in zedcloud/send.go. But hard to tell how much of an issue this is when we erroneously think COM ports should be used by TCP/IP. So it makes sense to merge this and keep on testing.

**Comment by andrewd-zededa** [comment]
I'm reviewing the eden test failures and here are the first three failures all in the Log/Info/Metric lim test, unable to ssh.  Not sure why yet

1. smoke ext4 true
        FAIL: ../lim/testdata/log_test.txt:16: command failure
        [background] bash ssh.sh: signal: killed
        [stdout]
        time="2024-12-11T21:11:25Z" level=fatal msg="command ssh failed: exit status 255"
        Connection timed out during banner exchange
        Connection to 127.0.0.1 port 2222 timed out
2. smoke ext4 false
        FAIL: ../lim/testdata/log_test.txt:16: command failure
        [background] bash ssh.sh: exit status 0
        [stdout]
        time="2024-12-11T21:11:30Z" level=fatal msg="command ssh failed: exit status 255"
        Connection timed out during banner exchange
        Connection to 127.0.0.1 port 2222 timed out
3. smoke zfs true
        FAIL: ../lim/testdata/log_test.txt:16: command failure
        [background] bash ssh.sh: exit status 0
        [stdout]
        time="2024-12-11T21:11:01Z" level=fatal msg="command ssh failed: exit status 255"
        Connection timed out during banner exchange
        Connection to 127.0.0.1 port 2222 timed out

**Comment by OhmSpectator** [comment]
@uncleDecart, do you have any ideas on why the test failed? =)

**Comment by uncleDecart** [comment]
> @uncleDecart, do you have any ideas on why the test failed? =)

@OhmSpectator, @andrewd-zededa does it run locally? Could be that EVE crashes and then you can't ssh

**Comment by eriknordmark** [comment]
FWIW I kicked off ztests on old of the old test machines in the lab and that seems to be fine.

But I also did a qemu run on my laptop using
make rootfs installer-raw run-installer-raw run-target
and after that virtual device is onboarded to the controller it gets stuck.

tail -F /run/diag.out shows among other things
ERROR: App test-ztest-ubunu uuid 91a07a2f-2606-4a50-8067-84a9300e21b9 state DOWNLOADING error: Found error in content tree Ztest-xenial-amd64 attached to volume test-ztest-ubunu_0_m_1: No IP management port addresses with cost <= 0

cat /run/nim/DeviceNetworkStatus/* |jq shows an error
      "LastError": "link not up for interface eth0 (down)",

And ip link has eth0 but not the actual hardware interface keth0 (eth0 is a bridge so we can attach e.g., switch network instances to eth0, while keth0 is the actual hardware device driver).

So something is seriously amiss here, even though I can update a running physical device to the image from this PR and it runs fine.

File: pkg/pillar/types/dns.go [hunk 4]
```diff
@@ -682,31 +685,38 @@ func getLocalAddrListImpl(dns DeviceNetworkStatus,
...
-		return true
-	}
-	return false
-}
-
-// IsL3Port checks if an interface name belongs to a port with SystemAdapter attached.
-func IsL3Port(dns DeviceNetworkStatus, ifname string) bool {
-	for _, us := range dns.Ports {
-		if us.IfName != ifname {
+		if us.USBAddr != usbaddr {
```
Comment (eriknordmark) [inline pkg/pillar/types/dns.go:693]: I wonder whether we need to treat empty strings as wildcards in some of these comparisons.

File: pkg/pillar/types/dns.go [hunk 5]
```diff
@@ -732,7 +751,7 @@ func GetPortCost(dns DeviceNetworkStatus, ifname string) uint8 {
 
 func GetPort(dns DeviceNetworkStatus, ifname string) *NetworkPortStatus {
 	for _, us := range dns.Ports {
-		if us.IfName != ifname {
+		if ifname != "" && us.IfName != ifname {
```
Comment (eriknordmark) [inline pkg/pillar/types/dns.go:754]: Do we have callers which pass an empty string as the ifname argument? If they do this will match on the first management port which might not be what we want.

File: pkg/pillar/types/dns.go (hunk 4 above)
Comment (eriknordmark) [reply pkg/pillar/types/dns.go:693]: Turns out that is at least required for PCIAddr in IsPort since the comparison is between the model aka SystemAdapter, which might have an empty PCIAddr and the discovered PCIAddr on the running system, which is not empty.

File: pkg/pillar/types/dns.go (hunk 5 above)
Comment (andrewd-zededa) [reply pkg/pillar/types/dns.go:754]: getLocalAddrListImpl() only calls it with a non empty ifname.

Could VerifyAllIntf() pass an empty string?

**Comment by andrewd-zededa** [comment]
@eriknordmark @uncleDecart @mikem-zed @milan-zededa @OhmSpectator @rene 

Maybe a simpler and quicker to test fix would be to guard against the ib.Type == types.IoCom

In the scope of domainmgr.go where it calls types.IsPort() change it to

```
		if types.IsPort(ctx.deviceNetworkStatus, ib.Ifname, ib.UsbAddr, ib.UsbProduct, ib.PciLong) && (ib.Type != types.IoCom){
			isPort = true
			keepInHost = true
		}
```

Or something similar.

File: pkg/pillar/types/dns.go [hunk 6]
```diff
@@ -682,31 +685,38 @@ func getLocalAddrListImpl(dns DeviceNetworkStatus,
 }
 
 // Check if an interface name is a port owned by nim
-func IsPort(dns DeviceNetworkStatus, ifname string) bool {
+func IsPort(dns DeviceNetworkStatus, ifname string, usbaddr string, usbprod string, pciaddr string) bool {
 	for _, us := range dns.Ports {
-		if us.IfName != ifname {
+		if ifname != "" && us.IfName != ifname {
```
Comment (eriknordmark) [inline pkg/pillar/types/dns.go:690]: Why this change? It causes the complete breakage since a lookup of e.g. COM1 or audio0 (which do not have an ifname by definition) will match the first item in dns.Ports. If that first item is a network port like eth0, then IsPort will incorrectly return true.

**Comment by OhmSpectator** [comment]
Do we still need this PR after merging #4482 ?
